xarray/core/dataset.py

```diff
@@ -8715,29 +8730,53 @@ def curvefit(
                 "in fitting on scalar data."
             )
 
+        # Check that initial guess and bounds only contain coordinates that are in preserved_dims
+        for param, guess in p0.items():
+            if isinstance(guess, DataArray):
+                unexpected = list(set(guess.dims) - set(preserved_dims))
+                if unexpected:
+                    raise ValueError(
+                        f"Initial guess for '{param}' has unexpected dimensions "
+                        f"{unexpected}. It should only have dimensions that are in data "
+                        f"dimensions {preserved_dims}."
+                    )
+        for param, (lb, ub) in bounds.items():
+            for label, bound in zip(["Lower", "Upper"], [lb, ub]):
+                if isinstance(bound, DataArray):
+                    unexpected = list(set(bound.dims) - set(preserved_dims))
+                    if unexpected:
+                        raise ValueError(
+                            f"{label} bound for '{param}' has unexpected dimensions "
+                            f"{unexpected}. It should only have dimensions that are in data "
+                            f"dimensions {preserved_dims}."
+                        )
+
         # Broadcast all coords with each other
         coords_ = broadcast(*coords_)
         coords_ = [
             coord.broadcast_like(self, exclude=preserved_dims) for coord in coords_
         ]
+        n_coords = len(coords_)
 
         params, func_args = _get_func_args(func, param_names)
         param_defaults, bounds_defaults = _initialize_curvefit_params(
             params, p0, bounds, func_args
         )
         n_params = len(params)
-        kwargs.setdefault("p0", [param_defaults[p] for p in params])
-        kwargs.setdefault(
-            "bounds",
-            [
-                [bounds_defaults[p][0] for p in params],
-                [bounds_defaults[p][1] for p in params],
-            ],
-        )
 
-        def _wrapper(Y, *coords_, **kwargs):
+        def _wrapper(Y, *args, **kwargs):
```

Comment:
I agree, but I'm not sure it's possible. The signature is `apply_ufunc(func, *args, ...)`, which only broadcasts each element in `*args`, so it doesn't work if the element itself is a list, like in the case of the initial guess. If `a`, `b` and `c` are `float | DataArray`, then we can only broadcast them like `f(a, b, c)`, and not like `f([a, b, c])`.

Let me know if you think the code could be made easier to follow somehow.

Reply:
Okay, I found a way to do it: if we convert the outer list to an additional dimension for `coords_` and to each of the params & bounds, and add that additional dimension to `input_core_dims`, then we can pass array-valued arguments to `_wrapper` with `apply_ufunc`. I'm not sure if it's more readable though.

You can see the implementation [here](https://github.com/mgunyho/xarray/blob/986fa1f47894e47b32187222312a046bde3d7786/xarray/core/dataset.py#L8758-L8820) (see [here](https://github.com/mgunyho/xarray/commit/986fa1f47894e47b32187222312a046bde3d7786) for the diff compared to this version, and [here](https://github.com/pydata/xarray/compare/main...mgunyho:xarray:curvefit-multidimensional-guess-split-wrapper-args?diff=split#diff-763e3002fd954d544b05858d8d138b828b66b6a2a0ae3cd58d2040a652f14638L8738) for compared to main (note: you have to click to the Files tab, I wasn't able to link straight to the line in the diff)).

It still needs some work:
- I now use `Dataset(...).to_array(dim_name)` to do the broadcasting + adding a new dimension, but is this a good way to do it? Another way I could think of was something like `concat([coord.expand_dims(new_dim=[i]) for i, coord in enumerate(coords_)], "new_dim")`. Both options are not very easy to understand, and probably have quite a bit of overhead.
- The dimension names `"param"` and `"coord"` are now hard-coded, so it will not work if the input array has these dimensions. Do you have a better suggestion? The output already has the hard-coded dimension "param", so it's not so bad.

If you think it's worth pursuing this approach, I can add that commit to this branch and we can discuss further.

Reply:
I think I prefer the current version in this PR.

I apologize for the noise.

Reply:
No worries. I think the alternative version could be OK, if there was a more explicit / clearer way of saying "for each element in this list, broadcast them to be the same shape, and then concatenate them along a new temporary axis", other than `Dataset(...).to_array()`. But yeah neither option is clearly better.